itemframework/src/item/item_scene.cpp

```diff
@@ -821,185 +822,250 @@ bool ItemScene::loadFromXml(QDomElement& dom)
+        qreal   const y    = element.attribute(YPosAttrTag).toDouble();
+        qint32  const id   = element.attribute(IdAttrTag).toUInt();
 
-    QHash<qint32, AbstractItem*> idMap;
+        itemPtr->setPos(x, y);
+        itemPtr->setName(name);
+        connect(itemPtr.data(), SIGNAL(changed()), this, SIGNAL(sceneRealChanged()));
 
-    while (!n.isNull()) {
-        progress = actchild * 100 / childs;
+        return {itemPtr.take(), id};
+    };
 
-        QDomElement e = n.toElement();
+    auto extractConnector = [this, shouldConnectIO](QDomElement const& element,
+            QHash<qint32, AbstractItem*> const& itemIds) -> Item_Connector* {
+
+        auto loadConnector = [this](QDomElement const& element, ItemInput* input, ItemOutput* output) {
+            auto connector = new Item_Connector(output, input);
+            connector->load_additional(element); // Load user modified path or custom routing
+            connector->do_update();
+            connect(connector, SIGNAL(changed()), this, SIGNAL(sceneRealChanged()));
+            return connector;
+        };
+
+        qint32  const sourceObject          = element.attribute(FromItemAttrTag).toInt();
+        qint32  const sourceOutputIndex     = element.attribute(FromIndexAttrTag).toInt();
+        qint32  const destinationObject     = element.attribute(ToItemAttrTag).toInt();
+        qint32  const destinationInputIndex = element.attribute(ToIndexAttrTag).toInt();
+        QString const typeName              = element.attribute(TransportTypeAttrTag);
+        qint32  const typeId                = QMetaType::type(typeName.toStdString().c_str());
+
+        if (!itemIds.contains(sourceObject) || !itemIds.contains(destinationObject)) {
+            qCritical() << "Couldn't find begin and end object of connector. Connector deleted.";
+            return nullptr;
+
+        } else if (typeId == QMetaType::UnknownType) {
+            qCritical() << "Unknown transport type \"" << typeName << "\"! Connector deleted.";
+            return nullptr;
+        }
 
-        if (!e.isNull()) {
-            if (e.tagName() == GraphicsItemTag) {
-                QString const type = e.attribute(TypeAttrTag);
+        auto sourceItem      = itemIds[sourceObject];
+        auto destinationItem = itemIds[destinationObject];
 
-                if (reportProgress)  {
-                    emit loadingProgress(progress, "Loading " + type); //show the exact name of what is currently loading
-                }
+        auto const outputs = sourceItem->outputs();
+        auto const inputs  = destinationItem->inputs();
 
-                actchild++;
-                AbstractItem* newItem = PluginManager::instance()->createInstance<AbstractItem>(type);
+        if (sourceOutputIndex < 0 || sourceOutputIndex > outputs.count() - 1) {
+            qCritical() << QString("Couldn't find ouput for connector on item %1. Connector deleted.").arg(sourceItem->name());
```

Comment:
I think we should include something like this somewhere:

http://stackoverflow.com/questions/9255887/stl-function-to-test-whether-a-value-is-within-some-range/9259994#9259994

`if (!interval(0, outputs.count()).contains(sourceOutputIndex) {`

would be better readable.